### PR TITLE
Packages/utils

### LINKED ISSUE #2693: Decouple schema validation from core
### Prerequisites

- [x] I have read the [documentation](https://react-jsonschema-form.readthedocs.io/);
- [ ] In the case of a bug report, I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [ ] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net/n1k0/f2y3fq7L/6/) or a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### Description

`ajv@6` has served this project well. The `6.*` line (and indeed the `7.*` line #2542) have reached end-of-life, and if new serious issues were discovered, may not be fixed. With the success of the decoupled, but co-maintained, themes it is perhaps time to consider co-developing multiple schema validators that meet different needs.

While client-side validation is an important part of the rjsf experience, and should be encouraged with a sensible default, picking a validator carries a number of trade-offs:

- overall size
- time to first load
- time to subsequent load
- time to validate
- arbitrary code execution
- JSON Schema drafts supported (the current `ajv@8` drops `draft4`)

For example, 
- `ajv@>7` makes pre-compiled [standalone validators](https://ajv.js.org/standalone.html) possible, which can be tree-shaken to avoid bringing the compiler
- rjsf can be used to _draw_ a form, with a given theme, and another technique entirely is used to validate it client-side

### Proposal

- [ ] move most of the logic in `validate.js` to a `AJV6Validator` class
  - [ ] define a new `ValidationError` to which an `AjvError` can be converted 
  - [ ] move these to a new package, `@rjsf/validator-ajv6`
    - [ ] refactor the boilerplate of this class and errors to a `NoOpValidator`
      - [ ] move the base to a new `@rjsf/validator-base` package, as a dependency of `@rjsf/core`
- [ ] add a new package `@rjsf/validator-ajv8` to track features
- [ ] add a `validateFormData` to the form props, somewhat like #1286, and, depending on when it lands:
  - ... in `@rjsf/core 3.*` 
    - use `AJV6Validator.validateFormData` default
  - ... in `@rjsf/core 4.*`
    - if available from a `peerDependency` on `@rjsf/validator-ajv8`
      - use a `AJV8Validator.validateFormData`
    - otherwise, 
      - fall back to `NoOpValidator`
      - `console.warn`

Once the existing behavior and test suite have been

### PR DESCRIPTION
**Pull request by heath-freenome**
### Reasons for making this change

Partially fixes #2693 by beginning the migration of the utils functions to their own package.

This first pass migrates all of the utils functions that do NOT require schema resolution into the new `@rjsf/utils` package.
- As part of this migration, all of the `@rjsf/core` types have been moved into `src/types.ts` and updated with improved and increased typings
- All of the migrated `utils` functions in this PR have been converted to typescript with 100% unit testing

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/master/CHANGELOG.md) with a description of the PR
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

File: packages/utils/src/deepEquals.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import isEqualWith from 'lodash/isEqualWith';
+
+/** Implements a deep equals using the `lodash.isEqualWith` function, that provides a customized comparator that
+ * assumes all functions are equivalent.
+ *
+ * NOTE: In @rjsf/core < 5.0, this was implemented by a modified copy of the https://github.com/othiym23/node-deeper
```
Comment (epicfaace) [inline packages/utils/src/deepEquals.ts:6]: rm this note because functionality hasn't changed

File: packages/utils/src/deepEquals.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * function, but since we are using elsewhere `lodash`, which has an isEqual that did the same thing, we decided to just
+ * customize it for the same use-case thereby maintaining less code.
+ *
+ * @param a - The first element to compare
+ * @param b - The second element to compare
+ */
+export default function deepEquals(a: any, b: any): boolean {
+  return isEqualWith(a, b, (obj: any, other: any) => {
+    if (typeof obj === 'function' || typeof other === 'function') {
```
Comment (epicfaace) [inline packages/utils/src/deepEquals.ts:15]: Consider changing this to &&

**Comment by heath-freenome** [comment]
@nickgros @epicfaace @jacqueswho I've updated all of the functions with documentation... Please give them a look over and add any feedback. I will merge it to the long-lived `rjsf-v5` once I get one approval

File: packages/utils/src/toConstant.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+import { RJSFSchema } from './types';
+
+export default function toConstant(schema: RJSFSchema) {
+  if (Array.isArray(schema.enum) && schema.enum.length === 1) {
+    return schema.enum[0];
+  } else if (schema.hasOwnProperty(CONST_NAME)) {
+    return schema.const;
+  } else {
+    throw new Error('schema cannot be inferred as a constant');
+  }
```
Comment (heath-freenome) [inline packages/utils/src/toConstant.ts:11]: if (x) return x
return y

**Pull request merged**